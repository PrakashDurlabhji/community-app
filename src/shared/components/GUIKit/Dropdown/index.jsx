/* eslint-disable jsx-a11y/label-has-for */
/**
 * Dropdown component.
 */
import React, { useState, useRef } from 'react';
import PT from 'prop-types';
import _ from 'lodash';
import ReactSelect from 'react-select';
import './style.scss';
import iconDown from 'assets/images/dropdown-arrow.png';
import { config } from 'topcoder-react-utils';


function Dropdown({
  options,
  label,
  required,
  placeholder,
  onChange,
  errorMsg,
  size,
}) {
  const [internalOptions, setInternalOptions] = useState(options);
  const selectedOption = _.find(internalOptions, { selected: true });
  const [focused, setFocused] = useState(false);
  const delayedOnChange = useRef(
    _.debounce((q, cb) => cb(q), config.GUIKIT.DEBOUNCE_ON_CHANGE_TIME),
  ).current;
  const sizeStyle = size === 'lg' ? 'lgSize' : 'xsSize';

  return (
    <div
      onFocusCapture={() => setFocused(true)}
      onBlurCapture={() => setFocused(false)}
      className="dropdownContainer"
      styleName={`container ${sizeStyle} ${selectedOption ? 'haveValue' : ''} ${
        errorMsg ? 'haveError' : ''
      } ${focused ? 'isFocused' : ''}`}
    >
      <div styleName="relative">
        <ReactSelect
          autosize={false}
          autoBlur
          options={internalOptions.map(o => ({ value: o.label, label: o.label }))}
          value={selectedOption ? selectedOption.label : null}
          onChange={(value) => {
            if (value) {
              const newOptions = internalOptions.map(o => ({
                selected: value.label === o.label,
                label: o.label,
              }));
              setInternalOptions(newOptions);
              delayedOnChange(
                _.cloneDeep(newOptions),
                onChange,
              );
            }
          }}
          placeholder={`${placeholder}${placeholder && required ? ' *' : ''}`}
          clearable={false}
        />
        <img width="15" height="9" styleName="iconDropdown" src={iconDown} alt="dropdown-arraow-icon" />
      </div>
      {label ? (
        <span styleName="label">
          {label}
          {required ? <span>&nbsp;*</span> : null}
        </span>
      ) : null}
      {errorMsg ? (<span styleName="errorMessage">{errorMsg}</span>) : null}
    </div>
  );
}

Dropdown.defaultProps = {
  placeholder: '',
  label: '',
  required: false,
  onChange: () => {},
  errorMsg: '',
  size: 'lg',
};

Dropdown.propTypes = {
  options: PT.arrayOf(
    PT.shape({
      label: PT.string.isRequired,
      selected: PT.bool.isRequired,
    }),
  ).isRequired,
  placeholder: PT.string,
  label: PT.string,
  required: PT.bool,
  onChange: PT.func,
  errorMsg: PT.string,
  size: PT.oneOf(['xs', 'lg']),
};

export default Dropdown;
