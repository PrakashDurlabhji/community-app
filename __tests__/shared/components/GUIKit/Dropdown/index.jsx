import React from 'react';
import Renderer from 'react-test-renderer/shallow';

import Dropdown from 'components/GUIKit/Dropdown';


const rnd = new Renderer();

it('Default render', () => {
  rnd.render((<Dropdown options={[{ label: 'dropdownLable', selected: true }]} />));
  expect(rnd.getRenderOutput()).toMatchSnapshot();
});
