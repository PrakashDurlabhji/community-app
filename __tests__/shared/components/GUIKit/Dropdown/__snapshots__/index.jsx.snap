// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Default render 1`] = `
<div
  className="dropdownContainer src-shared-components-GUIKit-Dropdown-___style__container___1qhm8 src-shared-components-GUIKit-Dropdown-___style__lgSize___1fYu0 src-shared-components-GUIKit-Dropdown-___style__haveValue___1b1qb"
  onBlurCapture={[Function]}
  onFocusCapture={[Function]}
>
  <div
    className="src-shared-components-GUIKit-Dropdown-___style__relative___1Yq0F"
  >
    <Select
      arrowRenderer={[Function]}
      autoBlur={true}
      autosize={false}
      backspaceRemoves={true}
      backspaceToRemoveMessage="Press backspace to remove {label}"
      clearAllText="Clear all"
      clearRenderer={[Function]}
      clearValueText="Clear value"
      clearable={false}
      closeOnSelect={true}
      deleteRemoves={true}
      delimiter=","
      disabled={false}
      escapeClearsValue={true}
      filterOptions={[Function]}
      ignoreAccents={true}
      ignoreCase={true}
      inputProps={Object {}}
      isLoading={false}
      joinValues={false}
      labelKey="label"
      matchPos="any"
      matchProp="any"
      menuBuffer={0}
      menuRenderer={[Function]}
      multi={false}
      noResultsText="No results found"
      onBlurResetsInput={true}
      onChange={[Function]}
      onCloseResetsInput={true}
      onSelectResetsInput={true}
      openOnClick={true}
      optionComponent={[Function]}
      options={
        Array [
          Object {
            "label": "dropdownLable",
            "value": "dropdownLable",
          },
        ]
      }
      pageSize={5}
      placeholder=""
      removeSelected={true}
      required={false}
      rtl={false}
      scrollMenuIntoView={true}
      searchable={true}
      simpleValue={false}
      tabSelectsValue={true}
      trimFilter={true}
      value="dropdownLable"
      valueComponent={[Function]}
      valueKey="value"
    />
    <img
      alt="dropdown-arraow-icon"
      className="src-shared-components-GUIKit-Dropdown-___style__iconDropdown___2bPox"
      height="9"
      src="/community-app-assets/images/e740c51bb099110dff2900fa4fc94d72.png"
      width="15"
    />
  </div>
</div>
`;
